File: fsharp/DateTime.fs
module MyDateTime

open Utils

type MyDateTime = 
    {
        Year   : string
        Month  : string
        Day    : string
        Hour   : string
        Minute : string
        Second : string
    }

let createRecord lst =
    match lst with
    | year :: month :: day :: hour :: minute :: second :: [] -> 
        { Year = year; Month = month; Day = day; Hour = hour; Minute = minute; Second = second }
    | _ -> 
        { Year = "2000"; Month = "Jan"; Day = "01"; Hour = "01"; Minute = "00"; Second = "00"}

let generateDateComponents() =
    let year = randomNumBetween 1970 2025
    let month = randomNumBetween 1 13
    let day = randomNumBetween 1 31
    let hour = randomNumBetween 1 13
    let minute = randomNumUnder 60
    let second = randomNumUnder 60
    let elements = [year; month; day; hour; minute; second] |> List.map string
    let newElements = 
        elements
        |> List.map (fun x ->
            if String.length x < 2 then
                "0" + x
            else
                x
        )
    createRecord newElements

let convertMonthStyle monthNum =
    match monthNum with
        | "01" -> "Jan"
        | "02" -> "Feb"
        | "03" -> "Mar"
        | "04" -> "Apr"
        | "05" -> "May"
        | "06" -> "Jun"
        | "07" -> "Jul"
        | "08" -> "Aug"
        | "09" -> "Sep"
        | "10" -> "Oct"
        | "11" -> "Nov"
        | _ -> "Dec"

let generateDefaultStyleDate (dateTime : MyDateTime) = 
    let {Year=year; Month=month; Day=day; Hour=hour; Minute=minute; Second=second} = dateTime
    $"%s{year}-%s{month}-%s{day}T%s{hour}:%s{minute}:%s{second}"

let generateHttpStyleDate (dateTime : MyDateTime)= 
    let {Year=year; Month=m; Day=day; Hour=hour; Minute=minute; Second=second} = dateTime
    let month = convertMonthStyle m
    $"[%s{day}/%s{month}/%s{year}:%s{hour}:%s{minute}:%s{second} +0000]"

File: golang-part2/pkgs/datetime/datetime.go
package datetime

import (
	"golang-part2/pkgs/utils"
	"strconv"
	"strings"
)

func GenerateDateComponents() []string {
	year := strconv.Itoa(utils.RandomNumBetween(1970, 2025))
	month := strconv.Itoa(utils.RandomNumBetween(1, 13))
	if len(month) < 2 {
		month = "0" + month
	}
	day := strconv.Itoa(utils.RandomNumBetween(1, 32))
	if len(day) < 2 {
		day = "0" + day
	}
	return []string{year, month, day}
}

func GenerateRandomTime() string {
	var times []string
	randHour := strconv.Itoa(utils.RandomNumBetween(1, 13))
	if len(randHour) < 2 {
		randHour = "0" + randHour
	}
	times = append(times, randHour)
	for i := 0; i < 2; i++ {
		rand := strconv.Itoa(utils.RandomNumUnder(60))
		if len(rand) < 2 {
			rand = "0" + rand
		}
		times = append(times, rand)
	}
	return strings.Join(times, ":")
}

func GenerateStandardStyleRandomDate() string {
	components := GenerateDateComponents()
	return strings.Join(components, "-")
}

func ConstructStandardDateTime(date, time string) string {
	components := []string{date, time}
	return strings.Join(components, "T")
}

func GenerateRequestStyleRandomDate() string {
	times := GenerateDateComponents()
	year, m, day := times[0], times[1], times[2]
	var month string
	switch m {
	case "01":
		month = "Jan"
	case "02":
		month = "Feb"
	case "03":
		month = "Mar"
	case "04":
		month = "Apr"
	case "05":
		month = "May"
	case "06":
		month = "Jun"
	case "07":
		month = "Jul"
	case "08":
		month = "Aug"
	case "09":
		month = "Sep"
	case "10":
		month = "Oct"
	case "11":
		month = "Nov"
	case "12":
		month = "Dec"
	}
	dateComponents := []string{day, month, year}
	return strings.Join(dateComponents, "/")
}

func ConstructRequestDateTime(date, time string) string {
	return "[" + date + ":" + time + " +0000]"
}
